### PR TITLE
Remove Roles via Admin

### PR DESCRIPTION
**Pull request by juliannatetreault**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
The "Definition of Done" for this RFC is:
> An admin can remove a role from a user successfully through the admin interface.

This PR adds the ability for admins to remove roles from users via the `/admin/users` dashboard. Admins have the ability to remove any role, other than the `super_admin` role, from any user other than themself. To accomplish this, I have done the following:
- Added a `#destroy` action to the `Admin::UsersController` that contains a conditional that prevents admins from removing `super_admin` roles and roles from themselves
- Added a `destroy` route to `routes.rb`
- Removed "Current Roles" from the `_activity.html.erb` partial and into its own partial, `_current_roles.html.erb` for design purposes
- Added a conditional to `_current_roles.html.erb` that conditionally renders the "X" removal button on any given admin user page
- Added tests around the removal of roles

## Related Tickets & Documents
Relates RFC [#53](https://github.com/forem/rfcs/pull/53)

## QA Instructions, Screenshots, Recordings
1️⃣  **To QA the removal of a base role from a user, please first ensure that you are an admin, and then do the following:**
- Navigate to `/admin/users/:id/edit`:
Screen Shot 2021-02-11 at 3 06 56 PM

- Click the "X" button next to the role of your choice. You should be prompted to confirm the removal of the role:

Screen Shot 2021-02-11 at 3 08 07 PM

- After confirming the role removal, you should then be redirected back to `admin/users/:id/edit` and shown a flash message stating that the removal has been successful:

Screen Shot 2021-02-11 at 3 08 41 PM

2️⃣  **To QA the removal of a `super_admin` role from a user, please first ensure that you are an admin, and then do the following:**
- Navigate to `/admin/users/:id/edit`:
Screen Shot 2021-02-04 at 3 52 10 PM
- Notice that there is no "X" button next to the `super_admin` role:
Screen Shot 2021-02-04 at 3 52 27 PM

3️⃣  **To QA the removal of a role from yourself, please first ensure that you are an admin, and then do the following:**
- Navigate to your `/admin/users/:id/edit` page:
Screen Shot 2021-02-04 at 3 45 07 PM

- Notice that there is no "X" button next to your roles:
Screen Shot 2021-02-04 at 3 46 01 PM

### UI accessibility concerns?
N/A

## Added tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## Added to documentation?

- [x] [Admin Guide](https://app.gitbook.com/@forem/s/forem-admin-guide/~/drafts/-MSj5dlCo5-KeF5F3Eh5/admin/users/user-roles)
**_Heads Up_**: I have not merged my changes to the Admin Guide yet (see "Admin Guide" link) -- please feel free to take a look at my addition to the docs! 👀 
- [ ] README
- [ ] No documentation needed

## [optional] Are there any post-deployment tasks we need to perform?
N/A

## [optional] What gif best describes this PR or how it makes you feel?

Panda rolling down a hill

### CODE REVIEW

File: app/views/admin/users/_current_roles.html.erb
Comment (citizen428) [inline app/views/admin/users/_current_roles.html.erb:13]: **nitpick (non-blocking):** Maybe we can look into formatting this a bit differently, the content of this line starts _past_ what is supposed to be our maximum line length 😅

File: spec/requests/admin/users_manage_spec.rb
Comment (citizen428) [inline spec/requests/admin/users_manage_spec.rb:168]: In examples that contain more than one assertion we tend to [aggregate failures](https://relishapp.com/rspec/rspec-core/docs/expectation-framework-integration/aggregating-failures#use-%60:aggregate-failures%60-metadata) by using RSpec [metadata](https://relishapp.com/rspec/rspec-core/docs/expectation-framework-integration/aggregating-failures#use-%60:aggregate-failures%60-metadata):

```suggestion
    it "removes non-admin roles from non-super_admin users", :aggregate_failures do
```

File: spec/requests/admin/users_manage_spec.rb
Comment (citizen428) [inline spec/requests/admin/users_manage_spec.rb:179]: Same comment as above re metadata.

File: spec/requests/admin/users_manage_spec.rb
Comment (citizen428) [inline spec/requests/admin/users_manage_spec.rb:190]: Same comment as above re metadata.

File: app/controllers/admin/users_controller.rb
Comment (citizen428) [inline app/controllers/admin/users_controller.rb:45]: **nitpick (non-blocking):** It seems we only need an actual user instance in one location, in other cases `params[:user_id]` would be sufficient. We could avoid the database call unless absolutely necessary by using the id param everywhere and updating the `elsif` to this:

```ruby
elsif User.find(params[:user_id]).remove_role(role)
```

I'm not actually sure saving one potentially unnecessary DB call is worth the decreased readability though, but I think I'd at least move fetching the user to _after_ the `:super_admin` check.

File: app/views/admin/users/_current_roles.html.erb
Comment (citizen428) [inline app/views/admin/users/_current_roles.html.erb:7]: **suggestion (non-blocking):** We have some similar code in `app/views/admin/users/_activity.html.erb`. Maybe we could use a helper like `role_display_name` here? Something like

```ruby
role_display_name(role)
  return "suspended" if role.name == "banned"

  role.name
end
```

Then you cold change the code here to this:

```ruby
<% if role.name == "banned" || role.name == "super_admin" || current_user.id == @user.id %>
  <li><%= role.name.titlecase %></li>
```

But that's probably better done in a follow-up PR.

File: spec/requests/admin/users_manage_spec.rb
Comment (juliannatetreault) [reply spec/requests/admin/users_manage_spec.rb:168]: Great call!

File: app/views/admin/users/_current_roles.html.erb
Comment (juliannatetreault) [inline app/views/admin/users/_current_roles.html.erb:2]: I decided to pull "Current Roles" out into its own partial to make it more extensible and because with the addition of the removal button to roles, it no longer makes sense to show "Current Roles" on the `/admin/users/:id` page.

File: app/views/admin/users/_current_roles.html.erb
Comment (juliannatetreault) [inline app/views/admin/users/_current_roles.html.erb:15]: Open to suggestions on how to clean this up a bit if anyone has any! I tried to break this line up a bit to make it more readable and to help it fall within our target line length.

File: app/views/admin/users/edit.html.erb
Comment (juliannatetreault) [inline app/views/admin/users/edit.html.erb:17]: I decided to pull "Current Roles" out into its own partial to make it more extensible and because with the addition of the removal button to roles, it no longer makes sense to show "Current Roles" on the /admin/users/:id page.

File: app/views/admin/users/_current_roles.html.erb
Comment (juliannatetreault) [reply app/views/admin/users/_current_roles.html.erb:7]: Sounds good to me! Thanks so much for the suggestion @citizen428 😊

File: app/views/admin/users/_current_roles.html.erb
Comment (vaidehijoshi) [reply app/views/admin/users/_current_roles.html.erb:15]: A small nit pick—I know the RFC mentioned that the button should read “REMOVE”, but I find that very hard to read and a bit verbose when it repeats over and over the way we have it now. Perhaps just an “X” here instead, or perhaps we can use a icon that already exists in our app? I think this RFC was accepted before we had the concept of product weigh in/approval, otherwise I would’ve weighed in here sooner! But this should be an easy change 😅

Comment (juliannatetreault) [reply app/views/admin/users/_current_roles.html.erb]: @vaidehijoshi After looking through the rest of the admin dashboard, it looks like we use "Destroy" and "Remove" on the buttons used to remove badge achievements, HTML variants, navigation links, etc., so  I opted to go with an "X" in place of "REMOVE" for the role removal buttons per your suggestion!

File: app/controllers/admin/users_controller.rb
Comment (juliannatetreault) [inline app/controllers/admin/users_controller.rb:58]: I'm running into an issue where this feature is working a little _too_ well and rather than removing the particular `single_resource_admin` role with a click of the "X," it's removing _all_ `single_resource_admin` roles. 🙃 

That being said, within this `elsif` I'm trying to ensure that we _only_ remove the `single_resource_admin` role that is being clicked on to be removed. In order to remove a `single_resource_admin` role, we need to do the following:
```
User.remove_role(:single_resource_admin, RoleOfChoice)
```
My code is _almost_ working, in that it grabs the correct role, in almost the correct format, but where it is not working is that `resource_type` returns a string, but to properly remove a role, we need the instance of the resource rather than the string (as shown above in my code snippet). 

With all of this being said, I could use some help with the following:
1. Determining whether I'm approaching the removal of `single_resource_admin` roles correctly.
2. If it has been determined that I am approaching the removal of `single_resource_admin` roles correctly, then being pointed in the right direction for how I might adjust my code to return the `resource_type` in the necessary format, rather than as a string.
3. If it has been determined that I am not approaching the removal of `single_resource_admin` roles correctly, then being pointed in the right direction for how I might approach the removal of `single_resource_admin` roles differently.

I have looked at the Rolify docs, Forem docs, dev.to, SO, and other random QA sites for possible solutions for insight on removing a resource instance and have been unsuccessful in my searches. Additionally, I've tried removing the specific `single_resource_admin` role via `resource_name`, `resource_type`, and `resource_id` in various ways both in the `Admin::UsersController` and in the `_current_roles.html.erb` partial, with the help of my tests and `binding.pry`. Unfortunately, none of these attempts have gotten me quite where I need to be.

Any and all help is welcome and greatly appreciated at this point! 

cc @citizen428

Comment (citizen428) [reply app/controllers/admin/users_controller.rb:58]: @juliannatetreault I had written this long answer but forgot to publish the review 🙄 I'm unresolving this to make sure you see it.

Comment (juliannatetreault) [reply app/controllers/admin/users_controller.rb:58]: Thank you so much for taking a look and helping me out with this @citizen428! Everything is working as expected now. 🎉  I used `.safe_constantize` instead of `.constantize` since I read that it was better for security reasons. Code Climate doesn't seem happy with either of the two methods, though. 

In regard to extracting this out into a service: do you think that would warrant a follow-up PR or should it be done in this PR?

Comment (citizen428) [reply app/controllers/admin/users_controller.rb]: @juliannatetreault Good call on `safe_constantize` vs `constantize`. To be fair, the way we'd be using `constantize` here is not the pattern that leads to security issues (which is more along the lines of `something.constantize.new.method_call`), but it's a good habit to get into.

Re the service object, I'll leave it up to you. If you think it's beneficial to get this out faster, let's just deploy it and you do a follow-up.

File: app/controllers/admin/users_controller.rb
Comment (citizen428) [inline app/controllers/admin/users_controller.rb:44]: **suggestion (non-blocking):** Time for a little Rails trick 😃 Have you ever wondered how things like `Rails.env.production?` work? The answer is a helpful little class called [`ActiveSupport::StringInquirer`](https://api.rubyonrails.org/classes/ActiveSupport/StringInquirer.html). `String` also has an [inquiry method](https://www.rubydoc.info/gems/activesupport/String:inquiry) which turns any string into an instance of this class:

```ruby
role = params[:role].inquiry
role.single_resource_admin?
#=> false
role.super_admin?
#=> true
```

If you do this your conditions read nicer and you don't have to manually convert it back to a string by calling `to_s` in the flash message generation:

```ruby
role.class
#=> ActiveSupport::StringInquirer
role.humanize.titlecase
#=> "Super Admin"
```

File: app/controllers/admin/users_controller.rb
Comment (citizen428) [inline app/controllers/admin/users_controller.rb:57]: **problem (of sorts), blocking:** One thing to think about: `safe_constantize` potentially returns `nil` and `remove_role(:single_resource_admin, nil)` doesn't error will work but may not do what we want it to:

```ruby
user.roles.pluck(:name, :resource_type)
#=> [["super_admin", nil], ["single_resource_admin", "Config"]]
resouce_type = "oops"
user.remove_role(:single_resource_admin, resource_type.safe_constantize)
user.roles.pluck(:name, :resource_type)
#=> [["super_admin", nil]]
```

As you can see the `:single_resource_admin` role for `Config` is gone, even though the `resource_type` was non-existing. So we should probably add a check to make sure the result of `safe_constantize` isn't `nil`. I know that because of the way the form is built this shouldn't happen, but I think I'd still feel better if we address this.

**Review by citizen428 — CHANGES_REQUESTED** [review]
This is looking good Julianna! Just left one suggestion and one small change request.

File: app/controllers/admin/users_controller.rb
Comment (juliannatetreault) [reply app/controllers/admin/users_controller.rb:44]: This is so neat - TIL! 😄 The only issue that I'm running into is that when I change `role = params[:role].to_sym` to `role = params[:role].inquiry`, I run into an issue where _all_ `:single_resource_admin` roles are removed rather than the particular one that I'm trying to remove. I attempted to add `.to_sym` inline when removing a `:single_resource_admin` role, but it still removed all `single_resource_admin` roles rather than the particular one that I was trying to remove.

File: app/controllers/admin/users_controller.rb
Comment (juliannatetreault) [reply app/controllers/admin/users_controller.rb:57]: Done! ✔️

File: app/controllers/admin/users_controller.rb
Comment (citizen428) [reply app/controllers/admin/users_controller.rb:44]: Did you try to do that before or after adding the `nil` check?

Comment (juliannatetreault) [reply app/controllers/admin/users_controller.rb:44]: I _believe_ I tried it both before and after. I just tried using `.inquiry` in place of `.to_sym` again and ran into the same result. 😕

Comment (citizen428) [reply app/controllers/admin/users_controller.rb:44]: That's curious. It really isn't all that important, the code is fine as-is. But if you want I can take a look.

**Review by citizen428 — APPROVED** [review]

File: spec/requests/admin/users_manage_spec.rb
Comment (citizen428) [inline spec/requests/admin/users_manage_spec.rb:179]: Should we update this spec and make sure the user has two single resource rules first and that only one of them gets removed?

**Review by rhymes — APPROVED** [review]
Works well, good job @juliannatetreault!

File: spec/requests/admin/users_manage_spec.rb
Comment (juliannatetreault) [reply spec/requests/admin/users_manage_spec.rb:179]: Great suggestion - done! ✔️

File: app/controllers/admin/users_controller.rb
Comment (juliannatetreault) [reply app/controllers/admin/users_controller.rb]: Update: circling back around to extracting the `#destroy` action into a service per your suggestion @citizen428!

File: app/views/admin/users/_current_roles.html.erb
Comment (juliannatetreault) [reply app/views/admin/users/_current_roles.html.erb:7]: Circled back around to this in another PR - done! ✅

**Pull request merged**
